### PR TITLE
client: let GetTSAsync caller control the creation of tsoRequest

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

ref TiDB https://github.com/pingcap/tidb/pull/11846

current GetTSAsync will use sync.Pool to reuse tsoRequest.

and we saw it have some cost in sysbench point-get test:

image

In TiDB, it's better reuse request in session level to avoid contend.

### What is changed and how it works?

add new method with new parameter TsoReqAlloc, so caller can define how to reuse.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Add new api

Side effects

 - n/a

Related changes

 - n/a

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/pd/1702)

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -592,15 +627,18 @@ type TSFuture interface {
...
+func (req *TsoRequest) Wait() (physical int64, logical int64, err error) {
 	// If tso command duration is observed very high, the reason could be it
 	// takes too long for Wait() be called.
 	start := time.Now()
 	cmdDurationTSOAsyncWait.Observe(start.Sub(req.start).Seconds())
 	select {
 	case err = <-req.done:
 		err = errors.WithStack(err)
-		defer tsoReqPool.Put(req)
+		if req.alloc != nil {
```
Comment (disksing) [inline client/client.go]: seems always true.

File: client/client.go [hunk 2]
```diff
@@ -563,24 +567,55 @@ func (c *client) GetURLs() []string {
 	return c.urls
 }
 
+// TsoReqAlloc defines how client create and reuse tsoRequest.
+type TsoReqAlloc interface {
```
Comment (disksing) [inline client/client.go]: Too many interfaces 😢 AFAICS, all you want is to use different pool instances for allocating the object. How about export `NewPool` instead, and pass an extra `pool` argument?

File: client/client.go [hunk 3]
```diff
@@ -563,24 +567,43 @@ func (c *client) GetURLs() []string {
...
```
Comment (disksing) [inline client/client.go]: Is it never used?

Comment (lysu) [reply client/client.go]: oh..that my fault, fixed~

**Comment by disksing** [comment]
@ngaut IIt will only take effect after updating tidb vendor.

File: client/client.go [hunk 4]
```diff
@@ -42,6 +42,8 @@ type Client interface {
 	GetTS(ctx context.Context) (int64, int64, error)
 	// GetTSAsync gets a timestamp from PD, without block the caller.
 	GetTSAsync(ctx context.Context) TSFuture
+	// GetTSAsync gets a timestamp from PD, without block the caller and provided a tsoRequest pool.
+	GetTSAsyncWithPool(ctx context.Context, pool *sync.Pool) TSFuture
```
Comment (ngaut) [inline client/client.go]: Pool can be used as package scope, no need to pass it every quest, it can save more memory.

File: client/client.go [hunk 5]
```diff
@@ -91,12 +93,14 @@ func WithExcludeTombstone() GetStoreOption {
...
 
-type tsoRequest struct {
+// TsoRequest presents the request to get tso.
+type TsoRequest struct {
 	start    time.Time
 	ctx      context.Context
 	done     chan error
 	physical int64
 	logical  int64
+	pool     *sync.Pool
```
Comment (ngaut) [inline client/client.go]: Remove the pool reference.

File: client/client.go (hunk 4 above)
Comment (lysu) [reply client/client.go]: this just wants to test divide global pool to session pool to avoid lock cost just as the description said

Comment (lysu) [reply client/client.go]: but maybe sync.Pool is not very suitable(after a gc will lost pooled request), maybe better add TsoReqAlloc interface back and use previous self-define pool impl https://github.com/pingcap/tidb/pull/11846/commits/43621c2325bf9c8e5d635d73d9c28343d0c5f6f1#diff-928cb614ae092ed4c6f349f4c4c991e4R200 :laughing:

Comment (ngaut) [reply client/client.go]: If the lock is an issue, we should solved it in client package(may be by using multiple pool) and keep the interface simple.  The Pool implementation is very fast, it has a fast path for get operation internally.

Comment (disksing) [reply client/client.go]: @lysu Sounds good to me if it is really necessary.

Comment (disksing) [reply client/client.go]: @lysu Could you try try Pool performance of go1.13? They claim to have some improvements.

File: client/client.go [hunk 6]
```diff
@@ -563,24 +567,35 @@ func (c *client) GetURLs() []string {
 	return c.urls
 }
 
+// NewReqAlloc creates a new TsoRequest.
+// only TsoReqAlloc implementor need use this.
+func NewReqAlloc() *TsoRequest {
```
Comment (shafreeck) [inline client/client.go]: Why not name as `NewTsoRequest` or `TsoRequestAlloc`?

**Comment by zhouqiang-cl** [comment]
@ngaut PD is not support /bench currently. we will support it next week.

**Pull request closed without merging**